### PR TITLE
[ATen] Keep the fill_ kernel for zero_ on ROCm 7.0.x (hipMemsetAsync corrupts under graph capture) (#197377)

### PR DESCRIPTION
**Pull request by sreedal**
Summary:

`zero_` on CUDA was moved off the `fill_` kernel and onto `cudaMemsetAsync` in
#195303 (`df69a036b4c2`, "[ATen] Zero CUDA tensors with cudaMemsetAsync instead
of a fill kernel"), which added `zero_cuda_` and repointed the `CUDA` dispatch
key for `zero_` in `native_functions.yaml`. On affected ROCm versions that path
silently returns wrong data: `hipMemsetAsync` captured into a HIP graph leaves
stale bytes behind, so a buffer `zero_` reports as zeroed reads back non-zero.
No error, no NaN, correct shape and dtype, so nothing downstream notices.

This is a ROCm runtime regression, not a PyTorch one, and it is bounded:

| ROCm  | memset under graph capture |
|---|---|
| 6.4.2 | clean |
| 7.0.0 | broken |
| 7.1.0 | clean |

The fix landed in ROCm/clr `3038c4c3777eeb55eabfddee56807c6f74b1b77c`, and
pdesupinski confirmed the bug does not reproduce on 7.1.0. So the window is
`[7.0.0, 7.1.0)`: `zero_cuda_` takes the memset fast path everywhere else, and
inside the window falls back to `self.fill_(0)` -- what `zero_` did before
#195303. CUDA is untouched.

Credit to pdesupinski for reducing this to `hipMemsetAsync` plus graph capture
and for pinning the fix commit. The original diff gated all of ROCm because the
mechanism was unclear at the time.

How it was found, for context on why the symptom looked stranger than the bug:
an LLM inference server on MI350X returned garbage completions. A decode step
zeroes a scratch buffer every iteration; when the zeroing does not take effect
the sampler reads stale logits and emits a long run of token id 0. The
production hit rate tracked concurrency -- 16% of responses at 64 concurrent
requests, 41% at 512, 6% at 8192 -- which is why light traffic looked clean.
Concurrency is not part of the trigger though; it only changes how often the
captured graph is replayed and how much stale data is sitting in the buffer.
Graph capture is the trigger, and disabling capture while leaving
`torch.compile` on was enough to make the corruption disappear:

| configuration | corrupt responses |
|---|---|
| memset, `torch.compile` on, graphs captured | 3.03% (174/5737) |
| memset, `torch.compile` on, graphs disabled | 0.00% (0/4656) |
| memset, `torch.compile` off, graphs disabled | 0.00% (0/5956) |
| `fill_` kernel, `torch.compile` on, graphs captured | 0.00% (0/5782) |

Authored with the assistance of an AI coding agent.

Test Plan:
Unit tests, `test/test_torch.py`:

* `test_zero_dense_emits_memset` asserts that dense `zero_` emits a `Memset`
  event. Which path `zero_` takes on ROCm is now version dependent, and the
  test-side ROCm version helper (`common_cuda.ROCM_VERSION`) is major/minor
  only, so rather than duplicate the version window in Python and have it rot
  against the C++, the test is skipped on ROCm and its original CUDA assertion
  is restored unchanged.
* `test_zero_dense_inside_graph_capture` is new, and is the one that actually
  covers this bug on ROCm. It captures `zero_` into a graph, replays it, and
  checks the buffer really is zero -- an assertion about behaviour rather than
  about which kernel ran, so it holds on every backend and every ROCm version.
  It is skipped where graph capture is unavailable.

```
python test/test_torch.py -k test_zero_
```

Honest status: these have not been run on a ROCm host. The MI350X capacity
reachable from here runs a packaged serving build, not a PyTorch source tree,
so `test_torch.py` cannot be invoked against it directly. The end-to-end runs
below exercise the same code path on the same hardware, but that is not a green
test run and is not claimed as one.

CUDA is unaffected by construction: `!defined(USE_ROCM)` makes the gate
unconditionally true on a CUDA build, so `zero_cuda_` compiles identically to
before this diff, and the profiler assertion on CUDA is the pre-existing one.

Minimal reproducer, pure PyTorch -- no serving stack, no model, no concurrency:

```python
import torch

buf = torch.empty(512, 151936, dtype=torch.float32, device="cuda")
sentinel = torch.full_like(buf, 7.0)
bad = torch.zeros((), dtype=torch.long, device="cuda")

def step():
    buf.copy_(sentinel)
    buf.zero_()
    bad.add_(buf.ne(0).sum())

side = torch.cuda.Stream()
side.wait_stream(torch.cuda.current_stream())
with torch.cuda.stream(side):
    for _ in range(3):
        step()
    graph = torch.cuda.CUDAGraph()
    with torch.cuda.graph(graph):
        step()
torch.cuda.current_stream().wait_stream(side)

bad.zero_()
for _ in range(20000):
    graph.replay()
torch.cuda.synchronize()
print("stale elements:", bad.item())   # expect 0
```

End-to-end evidence, collected before the mechanism was narrowed. A 0.6B
decoder-only model served with vLLM on one 8-GPU MI350X host, driven with a
fixed 200-token prompt and 100-token output, stop tokens disabled so every
request decodes the full 100 tokens. A response counts as corrupt if it
contains a run of token id 0. Each row is a separate run against a freshly
started server; checkpoint, sampling settings and every other flag held
constant, only the named column varied.

| concurrency | before (memset) | after (fill kernel) |
|---|---|---|
| 64 | 16.42% (33/201) | 0.00% (0/201) |
| 512 | 40.55% (339/836) | 0.00% (0/836) |
| 8192 | 6.31% (190/3012) | 0.00% (0/3000) |

Throughput and latency over the same runs, showing the fill kernel costs
nothing here:

| concurrency | | QPS | decode tok/s | avg latency | p50 TTFT | p99 TTFT |
|---|---|---|---|---|---|---|
| 64 | before | 89.09 | 8857 | 0.715s | 712.4ms | 824.1ms |
| 64 | after | 91.18 | 9041 | 0.699s | 700.4ms | 812.4ms |
| 512 | before | 561.44 | 55958 | 0.845s | 752.5ms | 2142.9ms |
| 512 | after | 566.42 | 56164 | 0.837s | 744.4ms | 2252.7ms |

`arc lint` on all four changed files: no issues.

Reviewed By: pdesupinski, atalman

Differential Revision: D120460087

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### CODE REVIEW

**Review by pdesupinski — APPROVED** [review]

**Comment by sreedal** [comment]
cc @ngimel for review - 

The previous pull request got reverted for mismatch even though the diffs were byte identical - assuming metadata mismatch. This is an exact copy of https://github.com/pytorch/pytorch/pull/197017 without any metadata modifications. Please help review.

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
Reach out to OSS team - bitwise identical diffs shouldn't be reverted
